vendor/vendor-android/base/src/main/kotlin/com/malinskiy/marathon/android/BaseAndroidDevice.kt

```diff
@@ -0,0 +1,224 @@
+package com.malinskiy.marathon.android
+
+import com.android.sdklib.AndroidVersion
+import com.malinskiy.marathon.analytics.internal.pub.Track
+import com.malinskiy.marathon.android.exception.InvalidSerialConfiguration
+import com.malinskiy.marathon.android.executor.listeners.CompositeTestRunListener
+import com.malinskiy.marathon.android.executor.listeners.DebugTestRunListener
+import com.malinskiy.marathon.android.executor.listeners.LogCatListener
+import com.malinskiy.marathon.android.executor.listeners.NoOpTestRunListener
+import com.malinskiy.marathon.android.executor.listeners.ProgressTestRunListener
+import com.malinskiy.marathon.android.executor.listeners.TestRunResultsListener
+import com.malinskiy.marathon.android.executor.listeners.screenshot.ScreenCapturerTestRunListener
+import com.malinskiy.marathon.android.executor.listeners.video.ScreenRecorderTestRunListener
+import com.malinskiy.marathon.android.serial.SerialStrategy
+import com.malinskiy.marathon.device.DeviceFeature
+import com.malinskiy.marathon.device.DevicePoolId
+import com.malinskiy.marathon.device.OperatingSystem
+import com.malinskiy.marathon.exceptions.DeviceSetupException
+import com.malinskiy.marathon.execution.Configuration
+import com.malinskiy.marathon.execution.TestBatchResults
+import com.malinskiy.marathon.execution.policy.ScreenRecordingPolicy
+import com.malinskiy.marathon.execution.progress.ProgressReporter
+import com.malinskiy.marathon.io.FileManager
+import com.malinskiy.marathon.log.MarathonLogging
+import com.malinskiy.marathon.report.attachment.AttachmentProvider
+import com.malinskiy.marathon.report.logs.LogWriter
+import com.malinskiy.marathon.test.TestBatch
+import com.malinskiy.marathon.time.Timer
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.runBlocking
+import java.util.*
+
+abstract class BaseAndroidDevice(
+    protected val adbSerial: String,
+    protected val serialStrategy: SerialStrategy,
+    protected val track: Track,
+    protected val timer: Timer
+) : AndroidDevice, CoroutineScope {
+    protected val logger = MarathonLogging.logger(AndroidDevice::class.java.simpleName)
+
+    override var abi: String = "Unknown"
+    override var version: AndroidVersion = AndroidVersion.DEFAULT
+    override var externalStorageMount: String = "Unknown"
```

Comment:
Why do you need override all fields in abstract class? + Why do you convert val -> var?
it was val in AndroidDevice.kt  and var here

Reply:
Because now they're essentially a lateinit var with a default value. The population of these happens during the setup call